### PR TITLE
Refactor tile coverage to support level of detail

### PR DESCRIPTION
**Pull request by mpulkki-mapbox**
This pull request implements a more versatile logic for finding visible tiles on the screen. Intersection tests have been extended to 3D space by using bounding volumes for tiles and camera instead of projecting everything to zero elevation plane. Tile pyramid is traversed depth-first while preserving distances to tiles allowing an easy support for a level-of-detail. Even though the distance based lod is supported by this implementation, **this PR has no visible effect for the user**. Level of detail logic is deactive when pitch is <= 60 degrees.

The native implementation follows the javascript one (https://github.com/mapbox/mapbox-gl-js/pull/8975) with few exceptions:
 1. Certain edge cases (false positives) were found in the tests. These cases (<1%) are culled by more precise intersection tests before reporting a tile to be visible.
 2. Few hotspots were found using a profiler so the order of intersection tests were slightly modified for a better performance.

Performance-wise the new implementation is few orders of magnitude slower than the original one. In a bigger picture the difference is not noticeable. I could see an increase of 5-15% time spent in TilePyramid::Update. Perhaps it's worth of discussion if the new tileCover should be a separate function and optional unless explicitly activated?

**TODO**
 - [x] Investigate north orientation of map

### CODE REVIEW

File: src/mbgl/util/tile_cover.cpp [hunk 1]
```diff
@@ -140,6 +141,128 @@ int32_t coveringZoomLevel(double zoom, style::SourceType type, uint16_t size) {
...
+    };
+
+    struct ResultTile {
+        OverscaledTileID id;
+        double sqrDist;
+    };
+
+    const double numTiles = std::pow(2.0, z);
+    const double worldSize = Projection::worldSize(state.getScale());
+    const uint8_t minZoom = state.getPitch() <= (60.0 / 180.0) * M_PI ? z : 0;
```
Comment (mpulkki-mapbox) [inline src/mbgl/util/tile_cover.cpp]: This line controls if tiles from different zoom levels can be returned. Level-of-detail functionality is disabled when pitch is <= 60 degrees.

File: src/mbgl/util/bounding_volumes.hpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    // Performs conservative intersection test using separating axis theorem.
+    // Some accuracy is traded for better performance. False positive rate is < 1%
+    IntersectionResult intersects(const AABB& aabb) const;
+
+    // Performs precise intersection test using separating axis theorem.
+    // It is possible run only edge cases that were not covered in intersects()
+    IntersectionResult intersectsPrecise(const AABB& aabb, bool edgeCasesOnly = false) const;
+
+    inline const std::array<vec3, 8>& getPoints() const { return points; }
```
Comment (pozdnyakov) [inline src/mbgl/util/bounding_volumes.hpp]: nit: `inline` is not needed

File: src/mbgl/util/bounding_volumes.hpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    inline const std::array<vec4, 6>& getPlanes() const { return planes; }
```
Comment (pozdnyakov) [inline src/mbgl/util/bounding_volumes.hpp]: ditto

File: src/mbgl/util/bounding_volumes.cpp [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+
+    if (flippedY) {
+        std::for_each(frustumPlanePointIndices.begin(), frustumPlanePointIndices.end(), [](vec3i& tri) {
+            std::swap(tri[1], tri[2]);
+        });
+    }
+
+    std::array<vec4, 6> frustumPlanes;
+
+    for (int i = 0; i < (int)frustumPlanePointIndices.size(); i++) {
```
Comment (pozdnyakov) [inline src/mbgl/util/bounding_volumes.cpp]: nit: ++i
why not `std::size_t`

**Comment by pozdnyakov** [comment]
Looks great, but let @astojilj to take a look before merge

**Comment by astojilj** [comment]
> Performance-wise the new implementation is few orders of magnitude slower than the original one. In a bigger picture the difference is not noticeable. I could see an increase of 5-15% time spent in TilePyramid::Update. Perhaps it's worth of discussion if the new tileCover should be a separate function and optional unless explicitly activated?

In case that this was measured using some benchmark, how much is "tile cover update" _budget_ per frame affected by this on example mobile platform?

File: src/mbgl/util/bounding_volumes.cpp [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+}
+
+IntersectionResult Frustum::intersectsPrecise(const AABB& aabb, bool edgeCasesOnly) const {
+    if (!edgeCasesOnly) {
+        IntersectionResult result = intersects(aabb);
+
+        if (result == IntersectionResult::Separate) return result;
+    }
+
+    const std::array<vec3, 4> aabbPoints = {vec3{aabb.min[0], aabb.min[1], 0.0},
```
Comment (astojilj) [inline src/mbgl/util/bounding_volumes.cpp]: Could alloc and copy here get removed, and add methods to access min and max also for dim 2?

Comment (mpulkki-mapbox) [reply src/mbgl/util/bounding_volumes.cpp]: All these arrays are stack allocated so the only cost should be variable initialization.

Z-dimension is redundant as all tiles are on the xy-plane and leaving it out reduces the number of required intersection tests. This could be updated if we ever have a need for elevation.

Corner points of the aabb could be pre-generated but it would be premature optimization in this case. Intersection tests are performed once per aabb and moving the generation of points to constructor wouldn't remove any logic. `Frustum::intersects` also requires a `w` component whereas `Frustum::intersectsPrecise` does not.

**Comment by astojilj** [comment]
LGTM after documenting performance impact on one frame.

**Comment by mpulkki-mapbox** [comment]
> > Performance-wise the new implementation is few orders of magnitude slower than the original one. In a bigger picture the difference is not noticeable. I could see an increase of 5-15% time spent in TilePyramid::Update. Perhaps it's worth of discussion if the new tileCover should be a separate function and optional unless explicitly activated?
> 
> In case that this was measured using some benchmark, how much is "tile cover update" _budget_ per frame affected by this on example mobile platform?

I profiled the code on my quite fast laptop in the following location: `[INFO] {mbgl-glfw}[General]: Exit location: --lat="60.169900" --lon="24.943500" --zoom="16.064100" --bearing "43.857600"`. I chose few test runs describing the average performance of both new and original tileCover implementation. These numbers are not exact but they give a **rough** idea of the performance impact.

original:
vtune_orig

new:
vtune_new

Tests were done by rendering a map 60fps for 10 seconds using "streets-v11" style. Time spent in `TilePyramid::update` increased ~10%. tileCover is also called from `render_background_layer.cpp`, but the profiler could not catch that. Frame budget used by `TilePyramid::update` during a 10s test run increases from ~2.7% to ~3.0% based on these observations. Impact on the overall tile budget is quite small.
